{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4a2c69a0",
   "metadata": {},
   "source": [
    "# This Notebook shows the basic usage and functionalities of metaKEGG.\n",
    "# Data displayed here are artificially generated. Ensemble IDs are arbitrary and do not reflect the official Gene Symbols. The pipelines can process human, mouse or any official gene symbols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "497bd9dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import metaKEGG\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8c4a37e",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "83e6ed27",
   "metadata": {},
   "outputs": [],
   "source": [
    "GENES_PATHWAYS_FILE_PATH = \"./single_input_genes.xlsx\"\n",
    "PATHWAYS_SHEET_NAME = \"pathways\"\n",
    "GENES_SHEET_NAME = \"gene_metrics\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7cf4b80",
   "metadata": {},
   "source": [
    "## Load Data and Inspect input file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c48334cc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Category",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Term",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "Count",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "%",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "PValue",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Genes",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "List Total",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Pop Hits",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Pop Total",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "Fold Enrichment Bonferroni",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Bonferroni",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "Benjamini",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "FDR",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "cc2d1d41-a08b-4113-a1d5-9646560f8372",
       "rows": [
        [
         "0",
         "KEGG_PATHWAY",
         "mmu04932:Non-alcoholic fatty liver disease",
         "130",
         "49.618320610687",
         "2.1538176e-06",
         "Ndufb4, Bcl2l11, Ndufb8, Atf4, Ndufa13, Il6, Il1a, Uqcrfs1, Bcl2l11, Prkag3, Xbp1, Prkag3, Ndufs8, Cox8c, Cox5a, Mapk14, Ndufb8, Adipor2, Ins2, Rxra, Tnf, Rela, Sdhb, Prkaa1, Tgfb1, Ndufb2, Pik3cb, Ndufb3, Cycs, Ndufs4, Il1a, Cox6b2, Prkaa1, Sdha, Rac1, Ndufb4b, Nr1h3, Cox7a2l, Cox6a2, Fasl, Mapk12, Jun, Cox8c, Mapk10, Ndufs1, Mapk11, COX3, Ndufa9, Ndufa7, Ndufv1, Bax, Ndufb10, CYTB, Ndufv3, Prkaa2, Pik3ca, Gsk3a, Ndufa3, Ndufa2, Tnfrsf1a, Ndufa5, Mlxip, Cox7b, Pklr, Cox7c, Cox8c, Ndufb9, Ppara, Cox7b2, Itch, Insr, Ndufs8, Srebf1, Uqcrq, Sdhd, Mapk14, Ndufb4b, Casp7, Eif2s1, Ppara, Rxra, Mapk14, Mapk11, Socs3, Pik3r3, Eif2s1, Ndufv2, Akt1, Pik3ca, Uqcr11, Bid, Ndufb6, Prkag2, Ndufv1, Akt3, Uqcr11, Il1b, Cox7b2, Prkag1, Bid, Adipoq, Casp7, Eif2ak3, Ndufa4l2, Ndufa5, Ndufa13, Nfkb1, Socs3, Cox5b, Ndufs2, Prkag2, Eif2ak3, Ppara, Ins2, Ndufa10, Adipor1, Bax, Mapk11, Ndufs1, Cox5a, Ndufs6, Ndufc1, Ndufs3, Prkag3, Akt3, Socs3, Cox7a2l, Ndufa1, COX1, Mapk13",
         "130",
         "157",
         "9396",
         "3.18567",
         "0.0129487",
         "0.0212391",
         "0.003745"
        ]
       ],
       "shape": {
        "columns": 13,
        "rows": 1
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Category</th>\n",
       "      <th>Term</th>\n",
       "      <th>Count</th>\n",
       "      <th>%</th>\n",
       "      <th>PValue</th>\n",
       "      <th>Genes</th>\n",
       "      <th>List Total</th>\n",
       "      <th>Pop Hits</th>\n",
       "      <th>Pop Total</th>\n",
       "      <th>Fold Enrichment Bonferroni</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Benjamini</th>\n",
       "      <th>FDR</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KEGG_PATHWAY</td>\n",
       "      <td>mmu04932:Non-alcoholic fatty liver disease</td>\n",
       "      <td>130</td>\n",
       "      <td>49.618321</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>Ndufb4, Bcl2l11, Ndufb8, Atf4, Ndufa13, Il6, I...</td>\n",
       "      <td>130</td>\n",
       "      <td>157</td>\n",
       "      <td>9396</td>\n",
       "      <td>3.18567</td>\n",
       "      <td>0.012949</td>\n",
       "      <td>0.021239</td>\n",
       "      <td>0.003745</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Category                                        Term  Count          %  \\\n",
       "0  KEGG_PATHWAY  mmu04932:Non-alcoholic fatty liver disease    130  49.618321   \n",
       "\n",
       "     PValue                                              Genes  List Total  \\\n",
       "0  0.000002  Ndufb4, Bcl2l11, Ndufb8, Atf4, Ndufa13, Il6, I...         130   \n",
       "\n",
       "   Pop Hits  Pop Total  Fold Enrichment Bonferroni  Bonferroni  Benjamini  \\\n",
       "0       157       9396                     3.18567    0.012949   0.021239   \n",
       "\n",
       "        FDR  \n",
       "0  0.003745  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathways_df = pd.read_excel(GENES_PATHWAYS_FILE_PATH , sheet_name=PATHWAYS_SHEET_NAME)\n",
    "pathways_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "85eb1d4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Category', 'Term', 'Count', '%', 'PValue', 'Genes', 'List Total',\n",
       "       'Pop Hits', 'Pop Total', 'Fold Enrichment Bonferroni', 'Bonferroni',\n",
       "       'Benjamini', 'FDR'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathways_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8f8019ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'mmu04932'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_ids = pathways_df['Term'].str.split(':')[0][0]\n",
    "pathway_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "56395b4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Non-alcoholic fatty liver disease'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_names = pathways_df['Term'].str.split(':')[0][1]\n",
    "pathway_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fc093cfa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Ndufb4', 'Bcl2l11', 'Ndufb8', 'Atf4', 'Ndufa13']"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_in_pathway = pathways_df['Genes'].str.split(', ')[0]\n",
    "genes_in_pathway[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "f0aefc51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "logFC",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "cf77ed08-9f35-4f45-adfb-6adbb0d24e4a",
       "rows": [
        [
         "0",
         "ENSMUSG00000000001",
         "Il6",
         "1.750551145434395"
        ],
        [
         "1",
         "ENSMUSG00000000002",
         "Il6ra",
         "2.086198899050252"
        ],
        [
         "2",
         "ENSMUSG00000000003",
         "Socs3",
         "-0.3062180431656762"
        ],
        [
         "3",
         "ENSMUSG00000000004",
         "Tnf",
         "-2.499700498424083"
        ],
        [
         "4",
         "ENSMUSG00000000005",
         "Tnfrsf1a",
         "2.71562584451735"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000000001</td>\n",
       "      <td>Il6</td>\n",
       "      <td>1.750551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000000002</td>\n",
       "      <td>Il6ra</td>\n",
       "      <td>2.086199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000000003</td>\n",
       "      <td>Socs3</td>\n",
       "      <td>-0.306218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000000004</td>\n",
       "      <td>Tnf</td>\n",
       "      <td>-2.499700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000000005</td>\n",
       "      <td>Tnfrsf1a</td>\n",
       "      <td>2.715626</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID gene_symbol     logFC\n",
       "0  ENSMUSG00000000001         Il6  1.750551\n",
       "1  ENSMUSG00000000002       Il6ra  2.086199\n",
       "2  ENSMUSG00000000003       Socs3 -0.306218\n",
       "3  ENSMUSG00000000004         Tnf -2.499700\n",
       "4  ENSMUSG00000000005    Tnfrsf1a  2.715626"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_metrics_df = pd.read_excel(GENES_PATHWAYS_FILE_PATH , sheet_name=GENES_SHEET_NAME)\n",
    "gene_metrics_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8ebce2f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_symbols = gene_metrics_df['gene_symbol'].tolist()\n",
    "logFCs = gene_metrics_df['logFC'].tolist()\n",
    "\n",
    "gene_symbol_logFC_dict = dict(zip(gene_symbols, logFCs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "37b14dca",
   "metadata": {},
   "source": [
    "## Run Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89aca17a",
   "metadata": {},
   "source": [
    "### Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "c4230a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a single input file as an argument and maps the detected genes on the enriched KEGG reference pathway, and colors them according to their log2FC values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ffa83b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "input_file_path = \"./single_input_genes.xlsx\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "count_threshold = 2\n",
    "pathway_pvalue_threshold = None\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "a0a73735",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis: Gene expression...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_genes\n",
      "New directory '.\\draw_KEGG_2025-12-03_genes' created  \n",
      "\n",
      "Output folder is .\\draw_KEGG_2025-12-03_genes\n",
      "\n",
      "Parsing input file...\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_genes\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column=genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps,\n",
    "                                folder_extension=folder_extension)\n",
    "\n",
    "# Run the pipeline\n",
    "my_pipeline.gene_expression(benjamini_threshold=benjamini_threshold,\n",
    "                            count_threshold=count_threshold,\n",
    "                            pathway_pvalue_threshold=pathway_pvalue_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ca4864d",
   "metadata": {},
   "source": [
    "### Transcript expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "23ba1cd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a single input file as an argument and maps the detected transcripts on the enriched KEGG reference pathway, and colors them according to their log2FC values.\n",
    "\n",
    "# NOTE: Pathway enrichement analysis with the DAVID Functional Annotation Tool, should be performed using transcript IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a3d9d627",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inspect the file\n",
    "\n",
    "TRANSCRIPTS_PATHWAYS_FILE_PATH = \"./single_input_transcripts.xlsx\"\n",
    "PATHWAYS_SHEET_NAME = \"pathways\"\n",
    "TRANSCRIPTS_SHEET_NAME = \"transcript_metrics\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "cfa4a84d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "logFC",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "5cfb954e-63af-43bf-ac9a-a2288f932c89",
       "rows": [
        [
         "0",
         "ENSMUST00000000001",
         "Il6",
         "0.4484760304930333"
        ],
        [
         "1",
         "ENSMUST00000000002",
         "Il6",
         "2.070120689250748"
        ],
        [
         "2",
         "ENSMUST00000000003",
         "Il6ra",
         "2.558086993709958"
        ],
        [
         "3",
         "ENSMUST00000000004",
         "Il6ra",
         "1.532811806847461"
        ],
        [
         "4",
         "ENSMUST00000000005",
         "Socs3",
         "2.645259515976938"
        ]
       ],
       "shape": {
        "columns": 3,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000000001</td>\n",
       "      <td>Il6</td>\n",
       "      <td>0.448476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000000002</td>\n",
       "      <td>Il6</td>\n",
       "      <td>2.070121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000000003</td>\n",
       "      <td>Il6ra</td>\n",
       "      <td>2.558087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000000004</td>\n",
       "      <td>Il6ra</td>\n",
       "      <td>1.532812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000000005</td>\n",
       "      <td>Socs3</td>\n",
       "      <td>2.645260</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ID gene_symbol     logFC\n",
       "0  ENSMUST00000000001         Il6  0.448476\n",
       "1  ENSMUST00000000002         Il6  2.070121\n",
       "2  ENSMUST00000000003       Il6ra  2.558087\n",
       "3  ENSMUST00000000004       Il6ra  1.532812\n",
       "4  ENSMUST00000000005       Socs3  2.645260"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_metrics_df = pd.read_excel(TRANSCRIPTS_PATHWAYS_FILE_PATH , sheet_name=TRANSCRIPTS_SHEET_NAME)\n",
    "transcript_metrics_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "04b645a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "count",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "6006cfb6-63b2-44d6-aace-5fc2dd42cb71",
       "rows": [
        [
         "Itch",
         "5"
        ],
        [
         "Cycs",
         "4"
        ],
        [
         "Fas",
         "4"
        ],
        [
         "Srebf1",
         "3"
        ],
        [
         "Nfkb1",
         "3"
        ]
       ],
       "shape": {
        "columns": 1,
        "rows": 5
       }
      },
      "text/plain": [
       "gene_symbol\n",
       "Itch      5\n",
       "Cycs      4\n",
       "Fas       4\n",
       "Srebf1    3\n",
       "Nfkb1     3\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_metrics_df['gene_symbol'].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "58b5f7b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "input_file_path = \"./single_input_transcripts.xlsx\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"transcript_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "count_threshold = 2\n",
    "pathway_pvalue_threshold = None\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "4fae2576",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis: Transcript expression...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_transcripts\n",
      "New directory '.\\draw_KEGG_2025-12-03_transcripts' created  \n",
      "\n",
      "Output folder is .\\draw_KEGG_2025-12-03_transcripts\n",
      "\n",
      "Parsing input file...\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_transcripts\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column=genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps,\n",
    "                                folder_extension=folder_extension)\n",
    "\n",
    "# Run the pipeline\n",
    "my_pipeline.transcript_expression(benjamini_threshold=benjamini_threshold,\n",
    "                                  count_threshold=count_threshold,\n",
    "                                  pathway_pvalue_threshold=pathway_pvalue_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe2f019",
   "metadata": {},
   "source": [
    "### Bulk RNAseq mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "7555c573",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes RANseq data, as single input file argument, maps the genes on a provided list of target pathways (assuming they are also found in the target pathways), and colors them according to their log2FC values."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "c729ef5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "input_file_path = \"./single_input_bulk.xlsx\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a1a38ef8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis : Bulk RNAseq mapping...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_bulk\n",
      "New directory '.\\draw_KEGG_2025-12-03_bulk' created  \n",
      "\n",
      "Output folder is .\\draw_KEGG_2025-12-03_bulk\n",
      "Parsing input file...\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_bulk\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column = genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps)\n",
    "# Run the pipeline\n",
    "my_pipeline.bulk_rnaseq_mapping()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5265c43",
   "metadata": {},
   "source": [
    "### Multiple inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "10fe9bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a list of inputs file as an argument and only maps pathways that are found in all of the inputs files.\n",
    "# For a common pathway, it will generate all possible states for a gene, from each individual input, to all possible combinations and assignes a unique color code to each combination.\n",
    "#     The detected genes are mapped enriched KEGG reference pathway, based on the state they're in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "154e8e0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input files list\n",
    "input_file_path = [\"./single_input_genes.xlsx\",\n",
    "                   \"./multiple_inputs_1.xlsx\",\n",
    "                   \"./multiple_inputs_2.xlsx\"]\n",
    "\n",
    "# Define parameters\n",
    "input_label = [\"input1\" , \"input2\" , \"input3\"]\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "count_threshold = 2\n",
    "pathway_pvalue_threshold = None\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c9d00c2f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis : Multiple inputs...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_3_inputs\n",
      "New directory '.\\draw_KEGG_2025-12-03_3_inputs' created  \n",
      "\n",
      "Output folder is .\\draw_KEGG_2025-12-03_3_inputs\n",
      "\n",
      "You want to map 3 inputs in total.\n",
      "Parsing input file...\n",
      "\n",
      "File Counter: 1, File: ./single_input_genes.xlsx, with name input1\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "File Counter: 2, File: ./multiple_inputs_1.xlsx, with name input2\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "File Counter: 3, File: ./multiple_inputs_2.xlsx, with name input3\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info & mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_3_inputs\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column=genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps)\n",
    "\n",
    "# Run the pipeline\n",
    "my_pipeline.multiple_inputs(input_label=input_label, # Label list for each input file\n",
    "                            count_threshold=count_threshold,\n",
    "                            benjamini_threshold=benjamini_threshold,\n",
    "                            pathway_pvalue_threshold=pathway_pvalue_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3bbefc3",
   "metadata": {},
   "source": [
    "### Differentially methylated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "979639e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a single input file and a methylation metadata file as arguments and maps the detected genes on the enriched KEGG reference pathway, and colors them according to their methylation state.\n",
    "# The state is defined as a binary reprsentation, depending if DMPs corresponding to a given gene are detected, or not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "3c4efaee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "CG_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "methylation_gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "methylation_pval",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "methylation_logFC",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "d3f2a2e5-7f0b-4f01-92ee-a1233492c866",
       "rows": [
        [
         "0",
         "cg_1",
         "Ndufs8",
         "0.0498008406105033",
         "1.915612075309638"
        ],
        [
         "1",
         "cg_2",
         "Uqcrq",
         "0.0137039582560795",
         "-0.8185928143597494"
        ],
        [
         "2",
         "cg_3",
         "Ndufb4b",
         "0.0400398499096648",
         "2.6488655412487763"
        ],
        [
         "3",
         "cg_4",
         "Cox5a",
         "0.0359382626979763",
         "2.818395984412035"
        ],
        [
         "4",
         "cg_5",
         "Nr1h3",
         "0.0305641404014649",
         "-0.2300531975484807"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CG_ID</th>\n",
       "      <th>methylation_gene_symbol</th>\n",
       "      <th>methylation_pval</th>\n",
       "      <th>methylation_logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cg_1</td>\n",
       "      <td>Ndufs8</td>\n",
       "      <td>0.049801</td>\n",
       "      <td>1.915612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cg_2</td>\n",
       "      <td>Uqcrq</td>\n",
       "      <td>0.013704</td>\n",
       "      <td>-0.818593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cg_3</td>\n",
       "      <td>Ndufb4b</td>\n",
       "      <td>0.040040</td>\n",
       "      <td>2.648866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cg_4</td>\n",
       "      <td>Cox5a</td>\n",
       "      <td>0.035938</td>\n",
       "      <td>2.818396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cg_5</td>\n",
       "      <td>Nr1h3</td>\n",
       "      <td>0.030564</td>\n",
       "      <td>-0.230053</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CG_ID methylation_gene_symbol  methylation_pval  methylation_logFC\n",
       "0  cg_1                  Ndufs8          0.049801           1.915612\n",
       "1  cg_2                   Uqcrq          0.013704          -0.818593\n",
       "2  cg_3                 Ndufb4b          0.040040           2.648866\n",
       "3  cg_4                   Cox5a          0.035938           2.818396\n",
       "4  cg_5                   Nr1h3          0.030564          -0.230053"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect the metadata file\n",
    "\n",
    "DMPS_FILE_PATH = \"./methylation.csv\"\n",
    "\n",
    "dmps_df = pd.read_csv(DMPS_FILE_PATH)\n",
    "dmps_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "e9eb2b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "input_file_path = \"./single_input_genes.xlsx\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "methylation_file_path = \"./methylation.csv\"\n",
    "methylation_genes_column = \"methylation_gene_symbol\"\n",
    "methylation_pvalue_column = \"methylation_pval\"\n",
    "methylation_pvalue_threshold = 0.05\n",
    "count_threshold = 2\n",
    "pathway_pvalue_threshold = None\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "14b4042f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis : Methylated genes...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_methylation\n",
      "New directory '.\\draw_KEGG_2025-12-03_methylation' created  \n",
      "\n",
      "Output folder is .\\draw_KEGG_2025-12-03_methylation\n",
      "\n",
      "Opening CSV file: ./methylation.csv\n",
      "Parsing input file...\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_methylation\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column=genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps)\n",
    "\n",
    "# Run the pipeline\n",
    "my_pipeline.methylated_genes(methylation_file_path=methylation_file_path,\n",
    "                              methylation_genes_column=methylation_genes_column,\n",
    "                              methylation_pvalue_column=methylation_pvalue_column,\n",
    "                              methylation_pvalue_threshold=methylation_pvalue_threshold,\n",
    "                              count_threshold=count_threshold,\n",
    "                              benjamini_threshold=benjamini_threshold,\n",
    "                              pathway_pvalue_threshold=pathway_pvalue_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f57095aa",
   "metadata": {},
   "source": [
    "### DMPs per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0a8d3e3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a single input file and a methylation metadata file as arguments and maps the detected genes on the enriched KEGG reference pathway.\n",
    "# It generates bins to quantify the number of DMPs that correspond to a given gene, and colors a gege according to its DMP bin.\n",
    "# The function also returns the quantification histogram plots, both in a grouped and an absolute count representation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "0a34d584",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "CG_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "methylation_gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "methylation_pval",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "methylation_logFC",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "7cc093b9-aad8-44d2-9179-f36de695a2d2",
       "rows": [
        [
         "0",
         "cg100_1",
         "Ndufs8",
         "0.049800841",
         "1.915612075"
        ],
        [
         "1",
         "cg100_2",
         "Uqcrq",
         "0.013703958",
         "-0.818592814"
        ],
        [
         "2",
         "cg100_3",
         "Ndufb4b",
         "0.04003985",
         "2.648865541"
        ],
        [
         "3",
         "cg100_4",
         "Cox5a",
         "0.035938263",
         "2.818395984"
        ],
        [
         "4",
         "cg100_5",
         "Nr1h3",
         "0.03056414",
         "-0.230053198"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CG_ID</th>\n",
       "      <th>methylation_gene_symbol</th>\n",
       "      <th>methylation_pval</th>\n",
       "      <th>methylation_logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cg100_1</td>\n",
       "      <td>Ndufs8</td>\n",
       "      <td>0.049801</td>\n",
       "      <td>1.915612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cg100_2</td>\n",
       "      <td>Uqcrq</td>\n",
       "      <td>0.013704</td>\n",
       "      <td>-0.818593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cg100_3</td>\n",
       "      <td>Ndufb4b</td>\n",
       "      <td>0.040040</td>\n",
       "      <td>2.648866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cg100_4</td>\n",
       "      <td>Cox5a</td>\n",
       "      <td>0.035938</td>\n",
       "      <td>2.818396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cg100_5</td>\n",
       "      <td>Nr1h3</td>\n",
       "      <td>0.030564</td>\n",
       "      <td>-0.230053</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     CG_ID methylation_gene_symbol  methylation_pval  methylation_logFC\n",
       "0  cg100_1                  Ndufs8          0.049801           1.915612\n",
       "1  cg100_2                   Uqcrq          0.013704          -0.818593\n",
       "2  cg100_3                 Ndufb4b          0.040040           2.648866\n",
       "3  cg100_4                   Cox5a          0.035938           2.818396\n",
       "4  cg100_5                   Nr1h3          0.030564          -0.230053"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect the metadata file\n",
    "\n",
    "DMPS_PER_GENE_FILE_PATH = \"./methylation_for_quantification.csv\"\n",
    "\n",
    "dmps_per_gene_df = pd.read_csv(DMPS_PER_GENE_FILE_PATH)\n",
    "dmps_per_gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "d02e6ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "methylation_gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "count",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "bcff3346-e37a-44a5-8988-fd84bdef0392",
       "rows": [
        [
         "Ppara",
         "13"
        ],
        [
         "Ndufs1",
         "10"
        ],
        [
         "Ndufb4b",
         "9"
        ],
        [
         "Bax",
         "9"
        ],
        [
         "Eif2s1",
         "8"
        ]
       ],
       "shape": {
        "columns": 1,
        "rows": 5
       }
      },
      "text/plain": [
       "methylation_gene_symbol\n",
       "Ppara      13\n",
       "Ndufs1     10\n",
       "Ndufb4b     9\n",
       "Bax         9\n",
       "Eif2s1      8\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dmps_per_gene_df['methylation_gene_symbol'].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "9fe7efac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "input_file_path = \"./single_input_genes.xlsx\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "methylation_file_path = \"./methylation_for_quantification.csv\"\n",
    "methylation_genes_column = \"methylation_gene_symbol\"\n",
    "methylation_pvalue_column = \"methylation_pval\"\n",
    "methylation_pvalue_threshold = 0.05\n",
    "methylation_probe_column = \"CG_ID\"\n",
    "probes_to_cgs=False\n",
    "count_threshold = 2\n",
    "pathway_pvalue_threshold = None\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "4da51b8e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis : DMPs per gene...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_methylation_quantification\n",
      "The contents of '.\\draw_KEGG_2025-12-03_methylation_quantification' have been removed, and the directory has been recreated.\n",
      "Output folder is .\\draw_KEGG_2025-12-03_methylation_quantification\n",
      "\n",
      "Opening CSV file: ./methylation_for_quantification.csv\n",
      "Will not perform probe correction...\n",
      "\n",
      "Parsing input file...\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_methylation_quantification\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column=genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps)\n",
    "\n",
    "# Run the pipeline\n",
    "my_pipeline.dmps_per_gene(methylation_file_path=methylation_file_path,\n",
    "                          methylation_genes_column=methylation_genes_column,\n",
    "                          methylation_pvalue_column=methylation_pvalue_column,\n",
    "                          methylation_pvalue_threshold=methylation_pvalue_threshold,\n",
    "                          methylation_probe_column=methylation_probe_column,\n",
    "                          probes_to_cgs=probes_to_cgs,\n",
    "                          count_threshold=count_threshold,\n",
    "                          benjamini_threshold=benjamini_threshold,\n",
    "                          pathway_pvalue_threshold=pathway_pvalue_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b98ec52",
   "metadata": {},
   "source": [
    "#### Probe correction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "7b9a3fa0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: When using probes_to_cgs=True, the pipeline will split the CG probes by the underscore '_' character and keep the first part, essentially correcting for different probe chemistry that could occur in the same position.\n",
    "\n",
    "# Example format is cg00000000_BC21 and cg00000000_TC21, which would be counted as two separate probes targeting the same gene.\n",
    "# Using the argument probes_to_cgs with True, the probes become cg00000000 & cg00000000, and duplicated entries per gene are eliminated, essentially counting one probe for the target gene."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf38a7eb",
   "metadata": {},
   "source": [
    "### miRNA target genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "db99a171",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a single input file and a miRNA metadata file as arguments and maps the detected genes on the enriched KEGG reference pathway, and colors them according to their miRNA state.\n",
    "# The state is defined as a binary reprsentation, depending if DEmiRs are targeting a given gene, or not."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "6e7637c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "miRNA_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "miRNA_gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "miRNA_pval",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "miRNA_logFC",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "4093b5f3-f944-4c18-b9fc-11404dc05e38",
       "rows": [
        [
         "0",
         "mmu-miR-1-3p",
         "dufb4",
         "0.5673016015041621",
         "1.2344330935433092"
        ],
        [
         "1",
         "mmu-miR-2-3p",
         "Atf4",
         "0.7567917920891399",
         "-0.0024237641218611"
        ],
        [
         "2",
         "mmu-miR-3-3p",
         "Prkag3",
         "0.7379026833863447",
         "2.747361383072688"
        ],
        [
         "3",
         "mmu-miR-4-3p",
         "Xbp1",
         "0.3816588055615935",
         "0.8020378157852837"
        ],
        [
         "4",
         "mmu-miR-5-3p",
         "Prkag3",
         "0.823298343695976",
         "2.2813966792063827"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>miRNA_gene_symbol</th>\n",
       "      <th>miRNA_pval</th>\n",
       "      <th>miRNA_logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mmu-miR-1-3p</td>\n",
       "      <td>dufb4</td>\n",
       "      <td>0.567302</td>\n",
       "      <td>1.234433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mmu-miR-2-3p</td>\n",
       "      <td>Atf4</td>\n",
       "      <td>0.756792</td>\n",
       "      <td>-0.002424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mmu-miR-3-3p</td>\n",
       "      <td>Prkag3</td>\n",
       "      <td>0.737903</td>\n",
       "      <td>2.747361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mmu-miR-4-3p</td>\n",
       "      <td>Xbp1</td>\n",
       "      <td>0.381659</td>\n",
       "      <td>0.802038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mmu-miR-5-3p</td>\n",
       "      <td>Prkag3</td>\n",
       "      <td>0.823298</td>\n",
       "      <td>2.281397</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       miRNA_ID miRNA_gene_symbol  miRNA_pval  miRNA_logFC\n",
       "0  mmu-miR-1-3p             dufb4    0.567302     1.234433\n",
       "1  mmu-miR-2-3p              Atf4    0.756792    -0.002424\n",
       "2  mmu-miR-3-3p            Prkag3    0.737903     2.747361\n",
       "3  mmu-miR-4-3p              Xbp1    0.381659     0.802038\n",
       "4  mmu-miR-5-3p            Prkag3    0.823298     2.281397"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DEMIRS_FILE_PATH = \"./miRNA.tsv\"\n",
    "\n",
    "demirs_df = pd.read_csv(DEMIRS_FILE_PATH , sep='\\t')\n",
    "demirs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "6658002a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "input_file_path = \"./single_input_genes.xlsx\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "miRNA_file_path = \"./miRNA.tsv\"\n",
    "miRNA_genes_column = \"miRNA_gene_symbol\"\n",
    "miRNA_pvalue_column = \"miRNA_pval\"\n",
    "miRNA_pvalue_threshold=0.05\n",
    "pathway_pvalue_threshold = None\n",
    "count_threshold = 2\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "17379d65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis : miRNA target genes...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_miRNA\n",
      "The contents of '.\\draw_KEGG_2025-12-03_miRNA' have been removed, and the directory has been recreated.\n",
      "Output folder is .\\draw_KEGG_2025-12-03_miRNA\n",
      "\n",
      "Opening TSV file: ./miRNA.tsv\n",
      "Parsing input file...\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_miRNA\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column=genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps)\n",
    "\n",
    "# Run the pipeline\n",
    "my_pipeline.mirna_target_genes(miRNA_file_path=miRNA_file_path,\n",
    "                                miRNA_genes_column=miRNA_genes_column,\n",
    "                                miRNA_pvalue_column=miRNA_pvalue_column,\n",
    "                                miRNA_pvalue_threshold=miRNA_pvalue_threshold,\n",
    "                                count_threshold=count_threshold,\n",
    "                                benjamini_threshold=benjamini_threshold,\n",
    "                                pathway_pvalue_threshold=pathway_pvalue_threshold)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6058c4fa",
   "metadata": {},
   "source": [
    "### DEmiRs per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3371a350",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This function takes a single input file and a miRNA metadata file as arguments and maps the detected genes on the enriched KEGG reference pathway.\n",
    "# It generates bins to quantify the number of DEmiRs that correspond to a given gene, and colors a gege according to its DEmiR bin.\n",
    "# The function also returns the quantification histogram plots, both in a grouped and an absolute count representation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "83cea0b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "miRNA_ID",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "miRNA_gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "miRNA_pval",
         "rawType": "float64",
         "type": "float"
        },
        {
         "name": "miRNA_logFC",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "ref": "4e0a7077-f490-4a81-b278-f01bd293b565",
       "rows": [
        [
         "0",
         "mmu-miR-1-3p",
         "dufb4",
         "0.5673016015041621",
         "1.2344330935433092"
        ],
        [
         "1",
         "mmu-miR-2-3p",
         "Atf4",
         "0.7567917920891399",
         "-0.0024237641218611"
        ],
        [
         "2",
         "mmu-miR-3-3p",
         "Prkag3",
         "0.7379026833863447",
         "2.747361383072688"
        ],
        [
         "3",
         "mmu-miR-4-3p",
         "Xbp1",
         "0.3816588055615935",
         "0.8020378157852837"
        ],
        [
         "4",
         "mmu-miR-5-3p",
         "Prkag3",
         "0.823298343695976",
         "2.2813966792063827"
        ]
       ],
       "shape": {
        "columns": 4,
        "rows": 5
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>miRNA_ID</th>\n",
       "      <th>miRNA_gene_symbol</th>\n",
       "      <th>miRNA_pval</th>\n",
       "      <th>miRNA_logFC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mmu-miR-1-3p</td>\n",
       "      <td>dufb4</td>\n",
       "      <td>0.567302</td>\n",
       "      <td>1.234433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mmu-miR-2-3p</td>\n",
       "      <td>Atf4</td>\n",
       "      <td>0.756792</td>\n",
       "      <td>-0.002424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mmu-miR-3-3p</td>\n",
       "      <td>Prkag3</td>\n",
       "      <td>0.737903</td>\n",
       "      <td>2.747361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mmu-miR-4-3p</td>\n",
       "      <td>Xbp1</td>\n",
       "      <td>0.381659</td>\n",
       "      <td>0.802038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mmu-miR-5-3p</td>\n",
       "      <td>Prkag3</td>\n",
       "      <td>0.823298</td>\n",
       "      <td>2.281397</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       miRNA_ID miRNA_gene_symbol  miRNA_pval  miRNA_logFC\n",
       "0  mmu-miR-1-3p             dufb4    0.567302     1.234433\n",
       "1  mmu-miR-2-3p              Atf4    0.756792    -0.002424\n",
       "2  mmu-miR-3-3p            Prkag3    0.737903     2.747361\n",
       "3  mmu-miR-4-3p              Xbp1    0.381659     0.802038\n",
       "4  mmu-miR-5-3p            Prkag3    0.823298     2.281397"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Inspect the metadata file\n",
    "\n",
    "DEMIRS_PER_GENE_FILE_PATH = \"./miRNA_for_quantification.tsv\"\n",
    "\n",
    "demirs_per_genes_df = pd.read_csv(DEMIRS_FILE_PATH , sep='\\t')\n",
    "demirs_per_genes_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "3abe6c9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "miRNA_gene_symbol",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "count",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "39e82ab2-5df6-43fe-9ff3-61c2e48a81a3",
       "rows": [
        [
         "Prkaa1",
         "2"
        ],
        [
         "Cox8c",
         "2"
        ],
        [
         "Ndufb8",
         "2"
        ],
        [
         "Il1a",
         "2"
        ],
        [
         "Bcl2l11",
         "2"
        ]
       ],
       "shape": {
        "columns": 1,
        "rows": 5
       }
      },
      "text/plain": [
       "miRNA_gene_symbol\n",
       "Prkaa1     2\n",
       "Cox8c      2\n",
       "Ndufb8     2\n",
       "Il1a       2\n",
       "Bcl2l11    2\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "demirs_per_genes_df['miRNA_gene_symbol'].value_counts().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "0cefadb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "\n",
    "input_file_path = \"./single_input_genes.xlsx\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "miRNA_file_path = \"./miRNA_for_quantification.tsv\"\n",
    "miRNA_genes_column = \"miRNA_gene_symbol\"\n",
    "miRNA_pvalue_column = \"miRNA_pval\"\n",
    "miRNA_pvalue_threshold = 0.05\n",
    "miRNA_column = \"miRNA_ID\"\n",
    "pathway_pvalue_threshold = None\n",
    "count_threshold = 2\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "b786c611",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis : DEmiRs per gene...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_miRNA_quantification\n",
      "New directory '.\\draw_KEGG_2025-12-03_miRNA_quantification' created  \n",
      "\n",
      "Output folder is .\\draw_KEGG_2025-12-03_miRNA_quantification\n",
      "\n",
      "Opening TSV file: ./miRNA_for_quantification.tsv\n",
      "Parsing input file...\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_miRNA_quantification\n"
     ]
    }
   ],
   "source": [
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                                sheet_name_paths=sheet_name_paths,\n",
    "                                sheet_name_genes=sheet_name_genes,\n",
    "                                genes_column=genes_column,\n",
    "                                log2fc_column=log2fc_column,\n",
    "                                save_to_eps=save_to_eps)\n",
    "\n",
    "my_pipeline.demirs_per_gene(miRNA_file_path=miRNA_file_path,\n",
    "                            miRNA_genes_column=miRNA_genes_column,\n",
    "                            miRNA_pvalue_column=miRNA_pvalue_column,\n",
    "                            miRNA_pvalue_threshold=miRNA_pvalue_threshold,\n",
    "                            miRNA_ID_column=miRNA_column, \n",
    "                            count_threshold=count_threshold,\n",
    "                            benjamini_threshold=benjamini_threshold,\n",
    "                            pathway_pvalue_threshold=pathway_pvalue_threshold)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96e71260",
   "metadata": {},
   "source": [
    "### Methylated + miRNA target genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "26655672",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define parameters\n",
    "input_file_path = \"./single_input_genes.xlsx\"\n",
    "sheet_name_paths = \"pathways\"\n",
    "sheet_name_genes = \"gene_metrics\"\n",
    "genes_column = \"gene_symbol\"\n",
    "log2fc_column = \"logFC\"\n",
    "methylation_file_path = \"./methylation.csv\"\n",
    "methylation_genes_column = \"methylation_gene_symbol\"\n",
    "methylation_pvalue_column = \"methylation_pval\"\n",
    "methylation_pvalue_threshold = 0.05\n",
    "miRNA_file_path = \"./miRNA.tsv\"\n",
    "miRNA_genes_column = \"miRNA_gene_symbol\"\n",
    "miRNA_pvalue_column = \"miRNA_pval\"\n",
    "miRNA_pvalue_threshold = 0.05\n",
    "count_threshold = 2\n",
    "pathway_pvalue_threshold = None\n",
    "benjamini_threshold = None\n",
    "save_to_eps = False\n",
    "folder_extension = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "e667f897",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing analysis : Methylated + miRNA target genes...\n",
      "\n",
      "Will create folder : .\\draw_KEGG_2025-12-03_methylation_and_miRNA\n",
      "The contents of '.\\draw_KEGG_2025-12-03_methylation_and_miRNA' have been removed, and the directory has been recreated.\n",
      "Output folder is .\\draw_KEGG_2025-12-03_methylation_and_miRNA\n",
      "\n",
      "Opening CSV file: ./methylation.csv\n",
      "Opening TSV file: ./miRNA.tsv\n",
      "Parsing input file...\n",
      "\n",
      "Will use thresholds:\n",
      "Minimum number of genes in pathway : 2 (included)\n",
      "pathway raw pvalue : None\n",
      "pathway Benjamini-Hochberg : None\n",
      "\n",
      "Finished parsing input file\n",
      "\n",
      "Collecting pathway info...\n",
      "Finished collecting pathway info\n",
      "\n",
      "Mapping pathways...\n",
      "\n",
      "[1/1] mmu04932 (Non-alcoholic fatty liver disease)\n",
      "Done! \n",
      "Output files are located in .\\draw_KEGG_2025-12-03_methylation_and_miRNA\n"
     ]
    }
   ],
   "source": [
    "# Intialize the object\n",
    "my_pipeline = metaKEGG.Pipeline(input_file_path=input_file_path,\n",
    "                        sheet_name_paths=sheet_name_paths,\n",
    "                        sheet_name_genes=sheet_name_genes,\n",
    "                        genes_column=genes_column,\n",
    "                        log2fc_column=log2fc_column,\n",
    "                        save_to_eps=save_to_eps)\n",
    "\n",
    "# Run the pipeline\n",
    "my_pipeline.methylated_and_mirna_target_genes(methylation_file_path=methylation_file_path,\n",
    "                                                    methylation_genes_column=methylation_genes_column,\n",
    "                                                    methylation_pvalue_column=methylation_pvalue_column,\n",
    "                                                    methylation_pvalue_threshold=methylation_pvalue_threshold,\n",
    "                                                    miRNA_file_path=miRNA_file_path,\n",
    "                                                    miRNA_genes_column=miRNA_genes_column,\n",
    "                                                    miRNA_pvalue_column=miRNA_pvalue_column,\n",
    "                                                    miRNA_pvalue_threshold=miRNA_pvalue_threshold,\n",
    "                                                    count_threshold=count_threshold,\n",
    "                                                    benjamini_threshold=benjamini_threshold,\n",
    "                                                    pathway_pvalue_threshold=pathway_pvalue_threshold)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "release",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
